### DISCUSSION TITLE
How to prioritize newer entries when searching?

### DISCUSSION THREAD

**Discussion by MileyHollenberg** [milvus-io/milvus#24972] [Q&A and General discussion]
I'm very new to Milvus and am currently trying to use it for an AI Assistant to give it long term memory. My project is in Typescript for Node and goes through a Discord bot to interface with it. It's mostly just a research project, nothing serious.

Currently I've added a bunch of test messages, "How's the weather?", and they all have a custom metadata field for the timestamp of when they were inserted. I'd like to be able to give a higher priority to the more recent messages in my search query but I don't know how to do that.

I'm able to sort the data after I've received a result but there's a high chance that the latest message won't even be in there and I don't think it's going to be a good idea to always try and get all data back and sort through it as that would cause my bot to become slower over time.

My current search function looks like this,

```
    public async findDocuments(collectionName: string, query: string, maxItems: number): Promise<Document[]> {
        const vectorStore = this.vectorStores.get(collectionName);
        if (!vectorStore) {
            throw new Error(`Collection ${collectionName} does not exist`);
        }

        const vectorsToSearch = await vectorStore.embeddings.embedQuery(query);

        // Specify a boolean expression for metadata filtering.
        // const expr = "timestamp == 1687073325056";

        // Perform the search with the metadata constraint.
        const searchResults = await this.milvusClient.search({
            collection_name: collectionName,
            data: vectorsToSearch,
            // expr,
            topk: maxItems,
            output_fields: ["langchain_text", "timestamp", "langchain_primaryid"],
        });

        let docs: Document[] = [];
        for (const result of searchResults.results) {
            docs.push({ pageContent: result.langchain_text, metadata: { timestamp: result.timestamp, id: result.id } });
        }

        docs = docs.sort((a, b) => {
            const aTimestamp = a.metadata.timestamp;
            const bTimestamp = b.metadata.timestamp;

            if (aTimestamp < bTimestamp) {
                return 1;
            }
            else if (aTimestamp > bTimestamp) {
                return -1;
            }
            else {
                return 0;
            }
        });

        return docs;
    }
```

Is there any way to make the search somehow prioritize more recent entries?

**Comment by xiaofan-luan** [comment]
Currently we will have to retrieve by score and sort by a field yourself.
I opened an issue to track

**Comment by yhmo** [comment]
Partitions might be a workaround.

Assume we want the AI to have "3 months of memory".
For each month we create a partition, messages of August in partition "Aug", messages of April in partition "April", messages of November in partition "Nov", etc.
In August, we only search the partitions "Aug", "July" and "June". In November, we only search the partitions "Nov", "Oct", "Step".
By this way, we can control the search scope to avoid the bot "become slower over time."

**Comment by ShravanSunder** [comment]
@xiaofan-luan how do you currently order by or rank in milvus?  I couldn't find much in the github or docs about ranking/reranking

**Reply by xiaofan-luan** [reply]
currently the answer is no, you could probably retrieve more than topk data and their timestamp and rerank by your self.
is this the same use case for agent?

**Reply by ShravanSunder** [reply]
@xiaofan-luan there's a couple of use cases.
- timestamp to make sure the memories loaded are within timeframe
- ranking by a certain field (say priority number or relevance pre-calculated number) to make sure its relevant to a agent's task.  
- order by pagination based on a field to process chunks of data.

**Comment by xiaofan-luan** [comment]
> @xiaofan-luan there's a couple of use cases.
> 
> * timestamp to make sure the memories loaded are within timeframe
> * ranking by a certain field (say priority number or relevance pre-calculated number) to make sure its relevant to a agent's task.
> * order by pagination based on a field to process chunks of data.

For timestamp frame a scalar filtering could work?
how to ranking between vectors and scalars is what should handle

**Comment by talhaanwarch** [comment]
This is much needed functionality.
